parso/tree.py

```diff
@@ -414,28 +415,19 @@ def get_leaf_for_position(self, position, include_prefixes=False):
             on whitespace or comments before a leaf
         :return: :py:class:`parso.tree.Leaf` at ``position``, or ``None``
         """
-        def binary_search(lower, upper):
-            if lower == upper:
-                element = self.children[lower]
-                if not include_prefixes and position < element.start_pos:
-                    # We're on a prefix.
-                    return None
-                # In case we have prefixes, a leaf always matches
-                try:
-                    return element.get_leaf_for_position(position, include_prefixes)
-                except AttributeError:
-                    return element
-
-            index = int((lower + upper) / 2)
-            element = self.children[index]
-            if position <= element.end_pos:
-                return binary_search(lower, index)
-            else:
-                return binary_search(index + 1, upper)
-
-        if not ((1, 0) <= position <= self.children[-1].end_pos):
+        if not ((1, 0) <= position < self.children[-1].end_pos):
             raise ValueError('Please provide a position that exists within this node.')
-        return binary_search(0, len(self.children) - 1)
+
+        index = bisect(self.children, position, key=lambda child: child.end_pos)
+        element = self.children[index]
+        if not include_prefixes and position < element.start_pos:
+            # We're on a prefix.
+            return None
+        # In case we have prefixes, a leaf always matches
+        try:
+            return element.get_leaf_for_position(position, include_prefixes)
+        except AttributeError:
```

Comment:
Sorry, but I don't want this. Why is this happening? Catching AttributeError in such a general fashion is something I never accept.

I'm fine with a non-recursive solution. Using bisect looks like a good idea. However this is something I really really want to avoid.

Reply:
this is just the existing code, moved

Reply:
Oh god. Doesn't make it better, but thanks for pointing it out.

Reply:
I guess you could instead ask whether the leaf is an instance of `BaseNode`, or even whether it has the `get_leaf_for_position` attribute.

The comment here seems confusing, I think what it really means is something like this:
```python
        # If we've found another node, recurse.
        if isinstance(element, BaseNode):
            return element.get_leaf_for_position(position, include_prefixes)

        return element
```

can make that change if you want - though I'm beginning to sense possible mission creep.  Let me know

Reply:
If it's really that easy I usually write it like this:

```
try:
    method = element.get_leaf_for_position
expect AttributeError:
    ...
else:
    return method(position, include_prefixes)
```

But let's first figure out the other problems.